linux/LinuxProcessList.c

```diff
@@ -486,6 +496,62 @@ static bool LinuxProcessList_readStatmFile(LinuxProcess* process, const char* di
    return (errno == 0);
 }
 
+static bool LinuxProcessList_readSmapsFile(LinuxProcess* process, const char* dirname, const char* name, bool haveSmapsRollup) {
+   //http://elixir.free-electrons.com/linux/v4.10/source/fs/proc/task_mmu.c#L719
+   //kernel will return data in chunks of size PAGE_SIZE or less.
+
+   char buffer[PAGE_SIZE];// 4k
+   char *start,*end;
+   ssize_t nread=0;
+   int tmp=0;
+   if(haveSmapsRollup) {// only available in Linux 4.14+
+      snprintf(buffer, MAX_NAME, "%s/%s/smaps_rollup", dirname, name);
+   } else {
+   snprintf(buffer, MAX_NAME, "%s/%s/smaps", dirname, name);
+   }
+   int fd = open(buffer, O_RDONLY);
+   if (fd == -1)
+      return false;
+
+   process->m_pss   = 0;
+   process->m_swap  = 0;
+   process->m_psswp = 0;
+
+   while ( ( nread =  read(fd,buffer, sizeof(buffer)) ) > 0 ){
+        start = (char *)&buffer;
+        end   = start + nread;
+        do{//parse 4k block
```

Comment:
I am afraid this loop won't read correctly the data from pages that fall in between two 4096-byte reads:

```
$ cat /proc/3304/task/3314/smaps | head -c 4096 | tail -n 10 
VmFlags: mr mw me 
2ec0a4f2b000-2ec0a4f3b000 r-xp 00000000 00:00 0 
Size:                 64 kB
Rss:                  64 kB
Pss:                  64 kB
Shared_Clean:          0 kB
Shared_Dirty:          0 kB
Private_Clean:         0 kB
Private_Dirty:        64 kB
Referenced:      %
```

For the above entry, it will read the Pss for this but not Swap and SwapPss, right?

I think we could get around the whole issue if we just add this feature for systems that have `smaps_rollup` available. Reading only the rollup version will simplify the patch and avoid the performance concerns.